tequilapi/http_api_server.go

```diff
@@ -0,0 +1,74 @@
+package tequilapi
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+)
+
+type ApiServer interface {
+	Port() int
+	Wait()
+	Stop()
+}
+
+type apiServer struct {
+	listener  net.Listener
+	stopped   *sync.WaitGroup
+	boundPort int
+}
+
+func NewServer(address string, port int, handler http.Handler) (ApiServer, error) {
+	var err error
+	listener, err := net.Listen("tcp", fmt.Sprintf("%s:%d", address, port))
+	if err != nil {
+		return nil, err
+	}
+
+	boundPort, err := extractBoundPort(listener)
```

Comment:
Why do we need port extracting? It's here  in `port` variable

Reply:
It covers a case when 0 port can be passed for listening, in that case OS will pickup random free port and bind on it. To avoid if port == 0 then do extract, this flow simply covers both scenarios.